Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,401 @@
+/-
+Copyright (c) 2024 Sihan Su. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca, Sihan Su, Wan Lin, Xiaoyang Su
+-/
+import Mathlib.Algebra.MvPolynomial.Monad
+import Mathlib.RingTheory.IntegralClosure.IsIntegralClosure.Basic
+/-!
+# Norther Normalizaton
+This file contains a proof of Noether normalization lemma by Nagata.
+
+## Main Results
+Let $A$ be a finitely generated algebra over a field $k$.
+Then there exists natural number $r$ and an injective homomorphism
+from $k[X_1, X_2, ..., X_r]$ to $A$ such that $A$ is integral over $k[X_1, X_2, ..., X_r]$.
+
+## Reference
+* <https://stacks.math.columbia.edu/tag/00OW>
+
+-/
+open Polynomial MvPolynomial Ideal BigOperators Nat RingHom
+
+variable {k : Type*} [Field k] {n : ℕ} (f : MvPolynomial (Fin (n + 1)) k)
+variable (v w : Fin (n + 1) →₀ ℕ)
+
+/-Suppose $f$ is a nonzero polynomial of $n+1$ variables : $X_0,...,X_n$.
+We construct a ring homomorphism $T$ which maps $X_i$ into $X_i + X_0^{r_i}$ when $i \neq 0$,
+and maps $f$ into some monic polynomial (regarded as a polynomial of $X_0$).-/
+private noncomputable abbrev up := (1 + ∑ i : Fin (n + 1), f.degreeOf i)
+
+private lemma up_spec (hv : v ∈ f.support) : ∀ i, v i < up f := by
+  intro i
+  have h : degreeOf i f ≤ ∑ i : Fin (n + 1), degreeOf i f := by
+    rw [← finsum_eq_sum_of_fintype fun i ↦ degreeOf i f]
+    exact single_le_finsum i (Set.toFinite _) (fun j ↦ Nat.zero_le (degreeOf j f))
+  exact lt_one_add_iff.mpr (le_trans (monomial_le_degreeOf i hv) h)
+
+private noncomputable abbrev r : Fin (n + 1) → ℕ :=
+    fun i ↦ if i = 0 then 1 else ((n + 1) * (up f)) ^ (n + 1 - i)
+
+/-$T$ is an algebraic isomorphism.-/
+private noncomputable abbrev T1 : MvPolynomial (Fin (n + 1)) k →ₐ[k] MvPolynomial (Fin (n + 1)) k :=
+  aeval (fun i ↦  if i = 0 then X 0 else X i + (X 0) ^ (r f i))
+
+private noncomputable abbrev T_inv :
+    MvPolynomial (Fin (n + 1)) k →ₐ[k] MvPolynomial (Fin (n + 1)) k :=
+  aeval (fun i ↦  if i = 0 then X 0 else X i - (X 0) ^ (r f i))
+
+private lemma T_left_inv : (T_inv f).comp (T1 f) = AlgHom.id _ _ := by
+  rw [comp_aeval, ← MvPolynomial.aeval_X_left]
+  congr
+  ext i v
+  cases' i using Fin.cases with d
+  · simp only [if_pos, MvPolynomial.aeval_X]
+  · simp only [Fin.succ_ne_zero, ite_false, map_add, map_pow,
+      MvPolynomial.aeval_X, if_pos, sub_add_cancel]
+
+private lemma T_right_inv : (T1 f).comp (T_inv f) = AlgHom.id _ _ := by
```

Comment:
You should be able to use `T_left_inv` to prove this ... if you generalize it a bit.
Generalize from `X 0 ^ r f i` to arbitrary polynomial in `X 0`.